src/track/beats.cpp

```diff
@@ -53,3 +190,666 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+    // a value of -1 indicates we went off the map -- count from the beginning.
+    double lower_bound = findNthBeatNew(curFrame, -n);
+    if (lower_bound == -1) {
+        lower_bound = m_beats.first().frame_position();
+    }
+
+    // If we hit the end of the beat map, recalculate the lower bound.
+    //double upper_bound = findNthBeat(lower_bound, n * 2);
+    double upper_bound = findNthBeatNew(lower_bound, n);
+    if (upper_bound == -1) {
+        upper_bound = m_beats.last().frame_position();
+        //lower_bound = findNthBeat(upper_bound, n * -2);
+        lower_bound = findNthBeatNew(upper_bound, n * -1);
+        // Super edge-case -- the track doesn't have n beats!  Do the best
+        // we can.
+        if (lower_bound == -1) {
+            lower_bound = m_beats.first().frame_position();
+        }
+    }
+
+    track::io::Beat startBeat, stopBeat;
+    startBeat.set_frame_position(lower_bound);
+    stopBeat.set_frame_position(upper_bound);
+    return calculateBpm(startBeat, stopBeat);
+}
+
+void Beats::addBeatNew(FrameNum beatFrame) {
+    QMutexLocker locker(&m_mutex);
+    track::io::Beat beat;
+    beat.set_frame_position(beatFrame);
+    BeatList::iterator it = std::lower_bound(
+            m_beats.begin(), m_beats.end(), beat, BeatLessThan);
+
+    // Don't insert a duplicate beat. TODO(XXX) determine what epsilon to
+    // consider a beat identical to another.
+    if (it->frame_position() == beat.frame_position())
+        return;
+
+    m_beats.insert(it, beat);
+    onBeatlistChanged();
+    locker.unlock();
+    emit updated();
+}
+
+void Beats::removeBeatNew(FrameNum beatFrame) {
+    QMutexLocker locker(&m_mutex);
+    track::io::Beat beat;
+    beat.set_frame_position(beatFrame);
+    BeatList::iterator it = std::lower_bound(
+            m_beats.begin(), m_beats.end(), beat, BeatLessThan);
+
+    // In case there are duplicates, remove every instance of dBeatSample
+    // TODO(XXX) add invariant checks against this
+    // TODO(XXX) determine what epsilon to consider a beat identical to another
+    while (it->frame_position() == beat.frame_position()) {
+        it = m_beats.erase(it);
+    }
+    onBeatlistChanged();
+    locker.unlock();
+    emit updated();
+}
+
+TimeSignature Beats::getSignatureNew(FrameNum frame) const {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid())
+        return kNullTimeSignature;
+
+    auto result = kDefaultTimeSignature;
+
+    // Special case, when looking for initial TimeSignature
+    if (frame == 0) {
+        auto beat = m_beats.cbegin();
+        if (beat->has_signature()) {
+            result.setBeats(beat->signature().beats());
+            result.setNoteValue(beat->signature().note_value());
+        }
+    } else {
+        // Scans the list of beats to find the last time signature change before the sample
+        for (auto beat = m_beats.begin(); beat != m_beats.end() && beat->frame_position() < frame; beat++) {
```

Comment:
I think this should search backwards starting from the frame position argument and `break` the loop when the first Beat with a TimeSignature is found.